File: examples/csharp/Customer.cs
// Copyright 2026 Chengxi Luo
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// `CustomerRecord` mirrors examples/cases/customer.yaml — a store account.
//
// `Address` and `Money` mirror the reusable address.yaml and money.yaml it
// imports; Order.cs reuses both, as the Go worker does.
//
// This is the only type in the suite carrying two formats, and the second one
// is the point: `binary` is a layout written by hand below, `json` goes through
// System.Text.Json, so the two fail in completely different ways. Both declare
// `oracle: semantic`, so what has to match the reference is the decoded value
// rather than the bytes — Go's encoder HTML-escapes `<`, `>` and `&` and always
// escapes U+2028/U+2029, and under a byte oracle every worker would have to
// reproduce that quirk.
//
// It is also the first C# model with nested structs. Nothing here says so: a
// property typed with another [SerifyModel] is a struct, an array of them a
// list<struct>, and a Dictionary of them a map<K,struct>.
//
// Go is the --ref language and owns the byte layout; see examples/go/wire.go.

using System;
using System.Buffers.Binary;
using System.Collections.Generic;
using System.IO;
using System.Text.Json;
using System.Text.Json.Nodes;
using Serify;

[SerifyModel]
internal sealed class Address
{
    [SerifyField] public string Recipient { get; set; } = "";
    [SerifyField] public string Street { get; set; } = "";
    [SerifyField] public string City { get; set; } = "";
    [SerifyField] public string Country { get; set; } = "";
    [SerifyField("postal_code")] public string PostalCode { get; set; } = "";

    /// <summary>A struct is its fields back to back, in schema order.</summary>
    public void Pack(MemoryStream ms)
    {
        Wire.WriteLenPrefixed(ms, Recipient);
        Wire.WriteLenPrefixed(ms, Street);
        Wire.WriteLenPrefixed(ms, City);
        Wire.WriteLenPrefixed(ms, Country);
        Wire.WriteLenPrefixed(ms, PostalCode);
    }

    public static Address Unpack(ReadOnlySpan<byte> d, ref int off) => new()
    {
        Recipient = Wire.ReadLenString(d, ref off),
        Street = Wire.ReadLenString(d, ref off),
        City = Wire.ReadLenString(d, ref off),
        Country = Wire.ReadLenString(d, ref off),
        PostalCode = Wire.ReadLenString(d, ref off),
    };

    public JsonObject ToJson() => new()
    {
        ["recipient"] = Recipient, ["street"] = Street, ["city"] = City,
        ["country"] = Country, ["postal_code"] = PostalCode,
    };

    public static Address FromJson(JsonNode? n) => new()
    {
        Recipient = (string?)n!["recipient"] ?? "",
        Street = (string?)n["street"] ?? "",
        City = (string?)n["city"] ?? "",
        Country = (string?)n["country"] ?? "",
        PostalCode = (string?)n["postal_code"] ?? "",
    };
}

[SerifyModel]
internal sealed class Money
{
    [SerifyField] public string Currency { get; set; } = "";
    [SerifyField("amount_minor")] public long AmountMinor { get; set; }

    public void Pack(MemoryStream ms)
    {
        Wire.WriteLenPrefixed(ms, Currency);
        Span<byte> buf = stackalloc byte[8];
        BinaryPrimitives.WriteInt64LittleEndian(buf, AmountMinor);
        ms.Write(buf);
    }

    public static Money Unpack(ReadOnlySpan<byte> d, ref int off)
    {
        var m = new Money { Currency = Wire.ReadLenString(d, ref off) };
        m.AmountMinor = BinaryPrimitives.ReadInt64LittleEndian(d[off..]);
        off += 8;
        return m;
    }

    public JsonObject ToJson() => new() { ["currency"] = Currency, ["amount_minor"] = AmountMinor };

    public static Money FromJson(JsonNode? n) => new()
    {
        Currency = (string?)n!["currency"] ?? "",
        AmountMinor = (long)n["amount_minor"]!,
    };
}

[SerifyModel]
internal sealed class CustomerRecord
{
    [SerifyField("customer_id")] public ulong CustomerId { get; set; }
    [SerifyField] public string Email { get; set; } = "";
    [SerifyField("display_name")] public string DisplayName { get; set; } = "";
    [SerifyField] public byte Age { get; set; }
    [SerifyField("email_verified")] public bool EmailVerified { get; set; }
    [SerifyField("fraud_score")] public float FraudScore { get; set; }
    [SerifyField("loyalty_points")] public uint LoyaltyPoints { get; set; }
    [SerifyField("signup_ts")] public long SignupTs { get; set; }
    [SerifyField("avatar_sha256")] public byte[] AvatarSha256 { get; set; } = Array.Empty<byte>();
    [SerifyField] public byte[] Pin { get; set; } = new byte[4];
    [SerifyField("referral_code")] public string? ReferralCode { get; set; }
    [SerifyField("store_credit")] public Money StoreCredit { get; set; } = new();
    [SerifyField("shipping_addresses")] public Address[] ShippingAddresses { get; set; } = Array.Empty<Address>();
    [SerifyField("address_book")] public Dictionary<string, Address> AddressBook { get; set; } = new();
    [SerifyField("wishlist_skus")] public string[] WishlistSkus { get; set; } = Array.Empty<string>();
    [SerifyField] public Dictionary<string, string> Preferences { get; set; } = new();

    private static void WriteCount(MemoryStream ms, int n)
    {
        Span<byte> buf = stackalloc byte[4];
        BinaryPrimitives.WriteUInt32LittleEndian(buf, (uint)n);
        ms.Write(buf);
    }

    public byte[] Marshal()
    {
        var ms = new MemoryStream();
        Span<byte> buf = stackalloc byte[8];

        BinaryPrimitives.WriteUInt64LittleEndian(buf, CustomerId);
        ms.Write(buf);
        Wire.WriteLenPrefixed(ms, Email);
        Wire.WriteLenPrefixed(ms, DisplayName);

        ms.WriteByte(Age);
        ms.WriteByte((byte)(EmailVerified ? 1 : 0));
        BinaryPrimitives.WriteSingleLittleEndian(buf, FraudScore);
        ms.Write(buf[..4]);
        BinaryPrimitives.WriteUInt32LittleEndian(buf, LoyaltyPoints);
        ms.Write(buf[..4]);
        BinaryPrimitives.WriteInt64LittleEndian(buf, SignupTs);
        ms.Write(buf);

        Wire.WriteLenPrefixed(ms, AvatarSha256);

        // array<T,N> carries no count: N is fixed by the schema.
        ms.Write(Pin);

        // optional<string>: a presence flag, then the value if present. An empty
        // string is present, which is why the flag cannot be inferred from it.
        if (ReferralCode is null) { ms.WriteByte(0); }
        else { ms.WriteByte(1); Wire.WriteLenPrefixed(ms, ReferralCode); }

        StoreCredit.Pack(ms);

        WriteCount(ms, ShippingAddresses.Length);
        foreach (var a in ShippingAddresses) a.Pack(ms);

        // Entry order is the dictionary's own — deliberately not sorted. A map
        // is unordered, so customer declares `oracle: semantic` and the decoded
        // value is what gets compared. See docs/protocol.md.
        WriteCount(ms, AddressBook.Count);
        foreach (var (k, a) in AddressBook) { Wire.WriteLenPrefixed(ms, k); a.Pack(ms); }

        WriteCount(ms, WishlistSkus.Length);
        foreach (var s in WishlistSkus) Wire.WriteLenPrefixed(ms, s);

        WriteCount(ms, Preferences.Count);
        foreach (var (k, v) in Preferences)
        {
            Wire.WriteLenPrefixed(ms, k);
            Wire.WriteLenPrefixed(ms, v);
        }

        return ms.ToArray();
    }

    public static CustomerRecord Unmarshal(byte[] data)
    {
        var c = new CustomerRecord();
        ReadOnlySpan<byte> d = data;
        var off = 0;

        c.CustomerId = BinaryPrimitives.ReadUInt64LittleEndian(d);
        off += 8;
        c.Email = Wire.ReadLenString(d, ref off);
        c.DisplayName = Wire.ReadLenString(d, ref off);

        c.Age = d[off];
        c.EmailVerified = d[off + 1] != 0;
        c.FraudScore = BinaryPrimitives.ReadSingleLittleEndian(d[(off + 2)..]);
        c.LoyaltyPoints = BinaryPrimitives.ReadUInt32LittleEndian(d[(off + 6)..]);
        c.SignupTs = BinaryPrimitives.ReadInt64LittleEndian(d[(off + 10)..]);
        off += 18;

        c.AvatarSha256 = Wire.ReadLenPrefixed(d, ref off).ToArray();

        c.Pin = d.Slice(off, 4).ToArray();
        off += 4;

        if (d[off++] == 0) { c.ReferralCode = null; }
        else { c.ReferralCode = Wire.ReadLenString(d, ref off); }

        c.StoreCredit = Money.Unpack(d, ref off);

        c.ShippingAddresses = new Address[ReadCount(d, ref off)];
        for (var i = 0; i < c.ShippingAddresses.Length; i++) c.ShippingAddresses[i] = Address.Unpack(d, ref off);

        c.AddressBook = new Dictionary<string, Address>();
        for (var n = ReadCount(d, ref off); n > 0; n--)
        {
            var k = Wire.ReadLenString(d, ref off);
            c.AddressBook[k] = Address.Unpack(d, ref off);
        }

        c.WishlistSkus = new string[ReadCount(d, ref off)];
        for (var i = 0; i < c.WishlistSkus.Length; i++) c.WishlistSkus[i] = Wire.ReadLenString(d, ref off);

        c.Preferences = new Dictionary<string, string>();
        for (var n = ReadCount(d, ref off); n > 0; n--)
        {
            var k = Wire.ReadLenString(d, ref off);
            c.Preferences[k] = Wire.ReadLenString(d, ref off);
        }

        return c;
    }

    private static int ReadCount(ReadOnlySpan<byte> d, ref int off)
    {
        var n = (int)BinaryPrimitives.ReadUInt32LittleEndian(d[off..]);
        off += 4;
        return n;
    }

    /// <summary>
    /// `bytes` is base64 and the 64-bit integers are JSON numbers: that is what
    /// the reference worker's `[]byte` and `uint64` marshal to, and the semantic
    /// oracle decodes our output with it.
    /// </summary>
    public byte[] ToJson()
    {
        var addrs = new JsonArray();
        foreach (var a in ShippingAddresses) addrs.Add(a.ToJson());

        var book = new JsonObject();
        foreach (var (k, a) in AddressBook) book[k] = a.ToJson();

        var skus = new JsonArray();
        foreach (var s in WishlistSkus) skus.Add(s);

        var prefs = new JsonObject();
        foreach (var (k, v) in Preferences) prefs[k] = v;

        var pin = new JsonArray();
        foreach (var b in Pin) pin.Add(b);

        var o = new JsonObject
        {
            ["customer_id"] = CustomerId,
            ["email"] = Email,
            ["display_name"] = DisplayName,
            ["age"] = Age,
            ["email_verified"] = EmailVerified,
            ["fraud_score"] = FraudScore,
            ["loyalty_points"] = LoyaltyPoints,
            ["signup_ts"] = SignupTs,
            ["avatar_sha256"] = Convert.ToBase64String(AvatarSha256),
            ["pin"] = pin,
            ["referral_code"] = ReferralCode,
            ["store_credit"] = StoreCredit.ToJson(),
            ["shipping_addresses"] = addrs,
            ["address_book"] = book,
            ["wishlist_skus"] = skus,
            ["preferences"] = prefs,
        };
        return JsonSerializer.SerializeToUtf8Bytes(o);
    }

    public static CustomerRecord FromJson(byte[] data)
    {
        var o = JsonNode.Parse(data)!;
        var c = new CustomerRecord
        {
            CustomerId = (ulong)o["customer_id"]!,
            Email = (string?)o["email"] ?? "",
            DisplayName = (string?)o["display_name"] ?? "",
            Age = (byte)o["age"]!,
            EmailVerified = (bool)o["email_verified"]!,
            FraudScore = (float)o["fraud_score"]!,
            LoyaltyPoints = (uint)o["loyalty_points"]!,
            SignupTs = (long)o["signup_ts"]!,
            AvatarSha256 = Convert.FromBase64String((string?)o["avatar_sha256"] ?? ""),
            ReferralCode = (string?)o["referral_code"],
            StoreCredit = Money.FromJson(o["store_credit"]),
        };

        var pin = o["pin"]!.AsArray();
        c.Pin = new byte[pin.Count];
        for (var i = 0; i < pin.Count; i++) c.Pin[i] = (byte)pin[i]!;

        var addrs = o["shipping_addresses"]!.AsArray();
        c.ShippingAddresses = new Address[addrs.Count];
        for (var i = 0; i < addrs.Count; i++) c.ShippingAddresses[i] = Address.FromJson(addrs[i]);

        foreach (var kv in o["address_book"]!.AsObject()) c.AddressBook[kv.Key] = Address.FromJson(kv.Value);

        var skus = o["wishlist_skus"]!.AsArray();
        c.WishlistSkus = new string[skus.Count];
        for (var i = 0; i < skus.Count; i++) c.WishlistSkus[i] = (string?)skus[i] ?? "";

        foreach (var kv in o["preferences"]!.AsObject()) c.Preferences[kv.Key] = (string?)kv.Value ?? "";

        return c;
    }
}

File: examples/php/customer.php
<?php
/*
 * Copyright 2026 Chengxi Luo
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * `CustomerRecord` mirrors examples/cases/customer.yaml — a store account.
 *
 * `Address` and `Money` mirror the reusable address.yaml and money.yaml it
 * imports; order.php reuses both, as the Go worker does.
 *
 * This is the only type in the suite carrying two formats, and the second one
 * is the point: `binary` is a layout written by hand below, `json` goes through
 * json_encode, so the two fail in completely different ways. Both declare
 * `oracle: semantic`, so what has to match is the decoded value rather than the
 * bytes — Go's encoder HTML-escapes `<`, `>` and `&` and always escapes
 * U+2028/U+2029, and under a byte oracle every worker would have to reproduce
 * that quirk.
 *
 * PHP's int is 64-bit *signed*, so a uint64 does not fit and `$customerId`
 * travels as a decimal string through ext-gmp, as the other models here do. The
 * same limit is what makes the JSON side awkward, since json_encode would then
 * quote it; see the two helpers at the bottom.
 *
 * It is also the first PHP model with nested structs. A property typed with
 * another #[SerifyModel] needs no declaration — its own type says which class
 * it is — but `array` says nothing about what it holds, so the list and map
 * properties name their element class with `elem:`.
 *
 * Go is the --ref language and owns the byte layout; see examples/go/wire.go.
 */

declare(strict_types=1);

use Serify\Attributes\SerifyField;
use Serify\Attributes\SerifyModel;

#[SerifyModel]
class Address
{
    #[SerifyField] public string $recipient = '';
    #[SerifyField] public string $street = '';
    #[SerifyField] public string $city = '';
    #[SerifyField] public string $country = '';
    #[SerifyField] public string $postalCode = '';

    /** A struct is its fields back to back, in schema order — nothing frames it. */
    public function pack(): string
    {
        return lenPrefixed($this->recipient) . lenPrefixed($this->street)
            . lenPrefixed($this->city) . lenPrefixed($this->country)
            . lenPrefixed($this->postalCode);
    }

    public static function unpack(string $data, int &$off): self
    {
        $a = new self();
        $a->recipient = readLenPrefixed($data, $off);
        $a->street = readLenPrefixed($data, $off);
        $a->city = readLenPrefixed($data, $off);
        $a->country = readLenPrefixed($data, $off);
        $a->postalCode = readLenPrefixed($data, $off);
        return $a;
    }

    public function toArray(): array
    {
        return [
            'recipient' => $this->recipient, 'street' => $this->street,
            'city' => $this->city, 'country' => $this->country,
            'postal_code' => $this->postalCode,
        ];
    }

    public static function fromArray(array $o): self
    {
        $a = new self();
        $a->recipient = $o['recipient'];
        $a->street = $o['street'];
        $a->city = $o['city'];
        $a->country = $o['country'];
        $a->postalCode = $o['postal_code'];
        return $a;
    }
}

#[SerifyModel]
class Money
{
    #[SerifyField] public string $currency = '';
    #[SerifyField] public string $amountMinor = '0';

    public function pack(): string
    {
        return lenPrefixed($this->currency) . encodeInt($this->amountMinor, 8);
    }

    public static function unpack(string $data, int &$off): self
    {
        $m = new self();
        $m->currency = readLenPrefixed($data, $off);
        $m->amountMinor = decodeSigned(substr($data, $off, 8));
        $off += 8;
        return $m;
    }

    public function toArray(): array
    {
        return ['currency' => $this->currency, 'amount_minor' => $this->amountMinor];
    }

    public static function fromArray(array $o): self
    {
        $m = new self();
        $m->currency = $o['currency'];
        $m->amountMinor = (string) $o['amount_minor'];
        return $m;
    }
}

#[SerifyModel]
class CustomerRecord
{
    #[SerifyField] public string $customerId = '0';
    #[SerifyField] public string $email = '';
    #[SerifyField] public string $displayName = '';
    #[SerifyField] public int $age = 0;
    #[SerifyField] public bool $emailVerified = false;
    #[SerifyField] public float $fraudScore = 0.0;
    #[SerifyField] public int $loyaltyPoints = 0;
    #[SerifyField] public string $signupTs = '0';
    #[SerifyField] public string $avatarSha256 = '';
    #[SerifyField] public array $pin = [];
    #[SerifyField] public ?string $referralCode = null;
    #[SerifyField] public Money $storeCredit;
    #[SerifyField(elem: Address::class)] public array $shippingAddresses = [];
    #[SerifyField(elem: Address::class)] public array $addressBook = [];
    #[SerifyField] public array $wishlistSkus = [];
    #[SerifyField] public array $preferences = [];

    public function __construct()
    {
        $this->storeCredit = new Money();
    }

    public function marshal(): string
    {
        $out = encodeInt($this->customerId, 8);
        $out .= lenPrefixed($this->email);
        $out .= lenPrefixed($this->displayName);
        $out .= pack('C', $this->age);
        $out .= pack('C', $this->emailVerified ? 1 : 0);
        $out .= pack('g', $this->fraudScore);
        $out .= pack('V', $this->loyaltyPoints);
        $out .= encodeInt($this->signupTs, 8);

        $out .= lenPrefixed($this->avatarSha256);

        // array<T,N> carries no count: N is fixed by the schema.
        foreach ($this->pin as $b) {
            $out .= pack('C', $b);
        }

        // optional<string>: a presence flag, then the value if present. An empty
        // string is present, which is why the flag cannot be inferred from it.
        if ($this->referralCode === null) {
            $out .= pack('C', 0);
        } else {
            $out .= pack('C', 1) . lenPrefixed($this->referralCode);
        }

        $out .= $this->storeCredit->pack();

        $out .= pack('V', count($this->shippingAddresses));
        foreach ($this->shippingAddresses as $a) {
            $out .= $a->pack();
        }

        // Entry order is the array's own — deliberately not sorted. A map is
        // unordered, so customer declares `oracle: semantic` and the decoded
        // value is what gets compared. See docs/protocol.md.
        $out .= pack('V', count($this->addressBook));
        foreach ($this->addressBook as $k => $a) {
            $out .= lenPrefixed((string) $k) . $a->pack();
        }

        $out .= pack('V', count($this->wishlistSkus));
        foreach ($this->wishlistSkus as $s) {
            $out .= lenPrefixed($s);
        }

        $out .= pack('V', count($this->preferences));
        foreach ($this->preferences as $k => $v) {
            $out .= lenPrefixed((string) $k) . lenPrefixed($v);
        }

        return $out;
    }

    public static function unmarshal(string $data): self
    {
        $c = new self();
        $off = 0;

        $c->customerId = decodeUnsigned(substr($data, 0, 8));
        $off = 8;
        $c->email = readLenPrefixed($data, $off);
        $c->displayName = readLenPrefixed($data, $off);

        $c->age = unpack('C', $data, $off)[1];
        $c->emailVerified = unpack('C', $data, $off + 1)[1] !== 0;
        $c->fraudScore = unpack('g', $data, $off + 2)[1];
        $c->loyaltyPoints = unpack('V', $data, $off + 6)[1];
        $c->signupTs = decodeSigned(substr($data, $off + 10, 8));
        $off += 18;

        $c->avatarSha256 = readLenPrefixed($data, $off);

        $c->pin = array_values(unpack('C4', substr($data, $off, 4)));
        $off += 4;

        if (unpack('C', $data, $off)[1] === 0) {
            $c->referralCode = null;
            $off += 1;
        } else {
            $off += 1;
            $c->referralCode = readLenPrefixed($data, $off);
        }

        $c->storeCredit = Money::unpack($data, $off);

        $c->shippingAddresses = [];
        for ($n = self::readCount($data, $off); $n > 0; $n--) {
            $c->shippingAddresses[] = Address::unpack($data, $off);
        }

        $c->addressBook = [];
        for ($n = self::readCount($data, $off); $n > 0; $n--) {
            $k = readLenPrefixed($data, $off);
            $c->addressBook[$k] = Address::unpack($data, $off);
        }

        $c->wishlistSkus = [];
        for ($n = self::readCount($data, $off); $n > 0; $n--) {
            $c->wishlistSkus[] = readLenPrefixed($data, $off);
        }

        $c->preferences = [];
        for ($n = self::readCount($data, $off); $n > 0; $n--) {
            $k = readLenPrefixed($data, $off);
            $c->preferences[$k] = readLenPrefixed($data, $off);
        }

        return $c;
    }

    private static function readCount(string $data, int &$off): int
    {
        $n = unpack('V', $data, $off)[1];
        $off += 4;
        return $n;
    }

    /**
     * `bytes` is base64 and the 64-bit integers are JSON numbers: that is what
     * the reference worker's `[]byte` and `uint64` marshal to, and the semantic
     * oracle decodes our output with it.
     */
    public function toJson(): string
    {
        $book = [];
        foreach ($this->addressBook as $k => $a) {
            $book[(string) $k] = $a->toArray();
        }

        $o = [
            'customer_id' => $this->customerId,
            'email' => $this->email,
            'display_name' => $this->displayName,
            'age' => $this->age,
            'email_verified' => $this->emailVerified,
            'fraud_score' => $this->fraudScore,
            'loyalty_points' => $this->loyaltyPoints,
            'signup_ts' => $this->signupTs,
            'avatar_sha256' => base64_encode($this->avatarSha256),
            'pin' => $this->pin,
            'referral_code' => $this->referralCode,
            'store_credit' => $this->storeCredit->toArray(),
            'shipping_addresses' => array_map(fn($a) => $a->toArray(), $this->shippingAddresses),
            // An empty PHP array encodes as [], not {}, and Go's decoder wants an
            // object here. JSON_FORCE_OBJECT would apply to every array in the
            // document, including the lists, so the cast is per-field.
            'address_book' => (object) $book,
            'wishlist_skus' => array_values($this->wishlistSkus),
            'preferences' => (object) $this->preferences,
        ];

        return unquoteBigints(json_encode($o, JSON_UNESCAPED_SLASHES | JSON_UNESCAPED_UNICODE));
    }

    public static function fromJson(string $data): self
    {
        $o = json_decode($data, true, 512, JSON_BIGINT_AS_STRING);
        $c = new self();

        // Every 64-bit field arrives as a string or an int depending on whether
        // it happened to fit in PHP's signed int; the model wants the decimal
        // string either way.
        $c->customerId = (string) $o['customer_id'];
        $c->email = $o['email'];
        $c->displayName = $o['display_name'];
        $c->age = $o['age'];
        $c->emailVerified = $o['email_verified'];
        // A JSON number is a double; narrow it the way the wire does, so a
        // float32 field holds a value float32 can actually represent.
        $c->fraudScore = unpack('g', pack('g', (float) $o['fraud_score']))[1];
        $c->loyaltyPoints = $o['loyalty_points'];
        $c->signupTs = (string) $o['signup_ts'];
        $c->avatarSha256 = base64_decode($o['avatar_sha256'] ?? '');
        $c->pin = $o['pin'];
        $c->referralCode = $o['referral_code'];
        $c->storeCredit = Money::fromArray($o['store_credit']);
        $c->shippingAddresses = array_map([Address::class, 'fromArray'], $o['shipping_addresses'] ?? []);
        $c->addressBook = array_map([Address::class, 'fromArray'], $o['address_book'] ?? []);
        $c->wishlistSkus = $o['wishlist_skus'] ?? [];
        $c->preferences = $o['preferences'] ?? [];

        return $c;
    }
}

// 64-bit integers in JSON
//
// PHP's int is signed, so max uint64 has no int form at all and json_encode
// would have to quote it. Reading is handled by JSON_BIGINT_AS_STRING; writing
// has no equivalent, so the three 64-bit fields are encoded as strings and
// unquoted in the text. They are the only 64-bit fields customer has.
//
// This rewrites the text rather than the value tree, so a *string value*
// spelling one of these keys followed by a colon and digits would be unquoted
// too. No case produces one, and the escape-hunting cases cannot: a `"` inside
// a JSON string is written `\"`.
const BIG_KEYS = 'customer_id|signup_ts|amount_minor';

function unquoteBigints(string $text): string
{
    return preg_replace('/"(' . BIG_KEYS . ')":"(-?\d+)"/', '"$1":$2', $text);
}
